Use more relevant test for TakeLast

Romasz opened the pull request:
I've played a little more with TakeLast method and looks like testing may be better - taking almost all the collection is not relevant for ICollection source, where we can take advantage of having the Count property. 
Here is a difference between taking half/full (Count-1) collection:

``` ini

BenchmarkDotNet=v0.11.3.1003-nightly, OS=Windows 10.0.17763.379 (1809/October2018Update/Redstone5)
Intel Core i7-7820HQ CPU 2.90GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.2.105
  [Host]     : .NET Core 2.1.9 (CoreCLR 4.6.27414.06, CoreFX 4.6.27415.01), 64bit RyuJIT
  Job-NUUCXS : .NET Core cd01132e-c21e-4542-a7b5-280fd76ce48a (CoreCLR 4.6.27513.77, CoreFX 4.7.19.17001), 64bit RyuJIT
  Job-OCUQKL : .NET Core de772c68-afb5-4ad9-ae49-71ca9083ee7f (CoreCLR 4.6.27513.77, CoreFX 4.7.19.17101), 64bit RyuJIT

Runtime=Core  IterationTime=250.0000 ms  MaxIterationCount=20  
MinIterationCount=15  WarmupCount=1  

```
|       Method |                                                                                                              Toolchain | size | iteration |            wrapType |       Mean |     Error |    StdDev |     Median |        Min |        Max | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------- |----------------------------------------------------------------------------------------------------------------------- |----- |---------- |-------------------- |-----------:|----------:|----------:|-----------:|-----------:|-----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
| **TakeLastFull** | **before** |  **100** |      **1000** |              **NoWrap** | **1,843.9 ns** | **34.345 ns** | **30.446 ns** | **1,837.5 ns** | **1,790.7 ns** | **1,896.1 ns** |  **1.00** |    **0.00** |      **0.3234** |           **-** |           **-** |              **1376 B** |
| TakeLastFull |        after |  100 |      1000 |              NoWrap |   985.8 ns |  6.878 ns |  6.434 ns |   985.0 ns |   976.1 ns |   998.3 ns |  0.54 |    0.01 |      0.0315 |           - |           - |               136 B |
|              |                                                                                                                        |      |           |                     |            |           |           |            |            |            |       |         |             |             |             |                     |
| TakeLastHalf | before |  100 |      1000 |              NoWrap | 1,614.0 ns | 26.925 ns | 25.185 ns | 1,615.0 ns | 1,571.6 ns | 1,664.3 ns |  1.00 |    0.00 |      0.1954 |           - |           - |               840 B |
| TakeLastHalf |        after |  100 |      1000 |              NoWrap |   512.3 ns |  4.349 ns |  3.632 ns |   511.9 ns |   507.6 ns |   518.0 ns |  0.32 |    0.01 |      0.0310 |           - |           - |               136 B |
|              |                                                                                                                        |      |           |                     |            |           |           |            |            |            |       |         |             |             |             |                     |
| **TakeLastFull** | **before** |  **100** |      **1000** |         **IEnumerable** | **1,784.7 ns** | **12.315 ns** | **11.520 ns** | **1,784.3 ns** | **1,766.0 ns** | **1,800.8 ns** |  **1.00** |    **0.00** |      **0.3343** |           **-** |           **-** |              **1400 B** |
| TakeLastFull |        after |  100 |      1000 |         IEnumerable | 1,902.3 ns | 37.360 ns | 34.946 ns | 1,908.0 ns | 1,834.4 ns | 1,970.8 ns |  1.07 |    0.02 |      0.3304 |           - |           - |              1400 B |
|              |                                                                                                                        |      |           |                     |            |           |           |            |            |            |       |         |             |             |             |                     |
| TakeLastHalf | before |  100 |      1000 |         IEnumerable | 1,606.3 ns | 26.728 ns | 22.319 ns | 1,599.4 ns | 1,578.9 ns | 1,654.2 ns |  1.00 |    0.00 |      0.2012 |           - |           - |               864 B |
| TakeLastHalf |        after |  100 |      1000 |         IEnumerable | 1,638.9 ns | 16.036 ns | 14.216 ns | 1,635.8 ns | 1,621.8 ns | 1,665.7 ns |  1.02 |    0.01 |      0.2030 |           - |           - |               864 B |
|              |                                                                                                                        |      |           |                     |            |           |           |            |            |            |       |         |             |             |             |                     |
| **TakeLastFull** | **before** |  **100** |      **1000** | **IReadOnlyCollection** | **1,963.1 ns** | **20.663 ns** | **19.328 ns** | **1,969.8 ns** | **1,921.8 ns** | **1,982.8 ns** |  **1.00** |    **0.00** |      **0.3308** |           **-** |           **-** |              **1400 B** |
| TakeLastFull |        after |  100 |      1000 | IReadOnlyCollection | 1,950.8 ns | 26.565 ns | 24.849 ns | 1,956.8 ns | 1,906.7 ns | 1,981.7 ns |  0.99 |    0.02 |      0.0552 |           - |           - |               256 B |
|              |                                                                                                                        |      |           |                     |            |           |           |            |            |            |       |         |             |             |             |                     |
| TakeLastHalf | before |  100 |      1000 | IReadOnlyCollection | 1,559.6 ns | 23.049 ns | 21.560 ns | 1,557.5 ns | 1,528.7 ns | 1,603.9 ns |  1.00 |    0.00 |      0.2033 |           - |           - |               864 B |
| TakeLastHalf |        after |  100 |      1000 | IReadOnlyCollection | 1,325.0 ns | 72.026 ns | 82.945 ns | 1,370.6 ns | 1,190.1 ns | 1,411.2 ns |  0.87 |    0.04 |      0.0587 |           - |           - |               256 B |
|              |                                                                                                                        |      |           |                     |            |           |           |            |            |            |       |         |             |             |             |                     |
| **TakeLastFull** | **before** |  **100** |      **1000** |         **ICollection** | **1,881.5 ns** | **10.885 ns** | **10.181 ns** | **1,882.5 ns** | **1,863.5 ns** | **1,894.7 ns** |  **1.00** |    **0.00** |      **0.3306** |           **-** |           **-** |              **1400 B** |
| TakeLastFull |        after |  100 |      1000 |         ICollection | 1,901.4 ns | 17.330 ns | 16.211 ns | 1,902.3 ns | 1,860.1 ns | 1,926.4 ns |  1.01 |    0.01 |      0.0537 |           - |           - |               256 B |
|              |                                                                                                                        |      |           |                     |            |           |           |            |            |            |       |         |             |             |             |                     |
| TakeLastHalf | before |  100 |      1000 |         ICollection | 1,544.1 ns | 22.777 ns | 21.306 ns | 1,535.6 ns | 1,514.6 ns | 1,579.1 ns |  1.00 |    0.00 |      0.2015 |           - |           - |               864 B |
| TakeLastHalf |        after |  100 |      1000 |         ICollection | 1,158.0 ns | 48.974 ns | 56.398 ns | 1,128.4 ns | 1,105.3 ns | 1,291.8 ns |  0.75 |    0.03 |      0.0583 |           - |           - |               256 B |

If you compare TakeLastFull for corerun before/after - there is no change in Mean time, but if you compare TakeLastHalf, you can see the difference.

maryamariyan commented:
Thanks @Romasz. Related to https://github.com/dotnet/corefx/pull/36051

maryamariyan commented:
Is this the only change that needs to be done to the performance repo? Don't we want to add new benchmarks for TakeLast?

cc: @VSadov @adamsitnik

Romasz commented:
Not sure  if the original test taking almost whole collection shouldn't stay - It may be worth to know the worst case if it hasn't slown down.

billwert commented:
I think having takefull/takehalf is great.

Romasz commented:
Here is a proposal of some changes:
  - added tests for One/Half/Full collection to obtain a full report,
  - added option to use multiple sizes for testing (https://github.com/dotnet/corefx/pull/36051#issuecomment-475623885)

adamsitnik commented on src/benchmarks/micro/corefx/System.Linq/Perf.Linq.cs:
```diff
@@ -107,54 +114,64 @@ public void Cast_SameType(int size, int iteration, Perf_LinqTestBase.WrapperType
+
+        [Benchmark]
+        [ArgumentsSource(nameof(IterationMultipleSizesWrapperData))]
         public void TakeLast(int size, int iteration, Perf_LinqTestBase.WrapperType wrapType)
```
to follow the convention you have introduced the method should be named TakeLast**Half**

```suggestion
        public void TakeLastHalf(int size, int iteration, Perf_LinqTestBase.WrapperType wrapType)
```

Romasz replied:
Yeah, this looks like a typo left from one of my numerous versions of the code.

Romasz commented:
@adamsitnik I completely agree regarding timings and number of performed tests. I just put this in the  pull request to know what you think about idea of providing _IterationMultipleSizesWrapperData_. From one POV it would be handy for other possible developers to have such construction by hand to perform multiple tests with varying collection size and compare in one summary. On the other hand I'm not sure how usefull it will be, hence probably there are not many methods requiring that. 

I've also some doubts regarding the implementation - the code depends on the order of variable instantiation which currently is little 'hidden'. I also thought about using [GlobalSetup] or static constructor to make it more readable. 

As for (100)x(TakeLastOne,TakeLastHalf,TakeLastFull)x(IEnumerable,ICollection) - IMHO it should be (100)x(TakeLastOne,TakeLastHalf,TakeLastFull)x(NoWrap,IEnumerable,ICollection) - NoWrap is IList which is a valuable information as currently System.Linq doesn't check for ICollection due to performance drop while checking for multiple interfaces.
